go/merkletree2/encoders.go

```diff
@@ -82,21 +118,47 @@ type SHA512_256Encoder struct{}
 
 var _ Encoder = SHA512_256Encoder{}
 
-func (e SHA512_256Encoder) EncodeAndHashGeneric(o interface{}) (Hash, error) {
-	enc, err := msgpack.EncodeCanonical(o)
+func (e SHA512_256Encoder) Encode(o interface{}) ([]byte, error) {
+	return msgpack.EncodeCanonical(o)
+}
+
+func (e SHA512_256Encoder) Decode(dest interface{}, src []byte) error {
+	return msgpack.Decode(dest, src)
+}
+
+func (e SHA512_256Encoder) EncodeAndHashGeneric(o interface{}) ([]byte, Hash, error) {
+	enc, err := e.Encode(o)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 	hasher := sha512.New512_256()
 	_, err = hasher.Write(enc)
+	if err != nil {
+		return nil, nil, err
+	}
+	return enc, hasher.Sum(nil), nil
+}
+
+func (e SHA512_256Encoder) HashKeyValuePairWithKeySpecificSecret(kvp KeyValuePair, kss KeySpecificSecret) (Hash, error) {
+	encVal, err := e.Encode(kvp.Value)
 	if err != nil {
 		return nil, err
 	}
-	return hasher.Sum(nil), nil
+	return e.HashKeyEncodedValuePairWithKeySpecificSecret(KeyEncodedValuePair{Key: kvp.Key, Value: encVal}, kss)
 }
 
-func (e SHA512_256Encoder) HashKeyValuePairWithKeySpecificSecret(kvp KeyValuePair, _ KeySpecificSecret) (Hash, error) {
-	return e.EncodeAndHashGeneric(kvp)
+func (e SHA512_256Encoder) HashKeyEncodedValuePairWithKeySpecificSecret(kevp KeyEncodedValuePair, kss KeySpecificSecret) (Hash, error) {
```

Comment:
`kss` isn't used, is that a bug?  Should it be using hmac?

Reply:
If this is a bug, let's make a new test to catch it.  Seems like it's not actually randomizing the tree?

Reply:
It's not a bug. This encoder does not perform any blinding by design (and is only used in tests) so it ignores kss. The blind tree will use the BlindedSHA512_256v1Encoder 
(which has a test ensuring that all values affect the output, in particular see encoders_test:71)

I agree that the silently ignoring blinding is misleading though. Maybe I should throw/panic if a non nil kss is being given as input to this encoder?

Reply:
sure